text/0079-cdk-v2.md

```diff
@@ -0,0 +1,339 @@
+bump.
+
+The following sections describe each one of the motivations above.
+
+## Dependency Management
+
+The CDK consists of 100+ packages, one for every AWS service and then some, with
+complex interdependencies. For example, the *aws-ecs* package depends on *core*,
+*aws-iam*, *aws-ecr*, *aws-ec2* and more packages to do its work. In fact, it
+depends on 23 other packages.
+
+This means that when a user wishes to use the *aws-ecs* module, their package
+manager needs to fetch all 23 dependencies.
+
+### Unacceptable experience with peer npm dependencies
+
+Most languages (Java, Python, .NET) only support a single instance of a module
+at runtime. This means that if two modules declare that they take a dependency
+on a third module, the dependency manager will resolve a common version and
+fetch it.
+
+Contrary to most languages, the Node.js ecosystem supports the co-existence of
+multiple versions of a module at runtime (since `node_modules` first searched as
+a relative path). This is problematic for the CDK case since CDK modules often
+expose types from other modules in their API (e.g. an `s3.Bucket` accepts a
+`kms.Key` as an option).
+
+To model this type of dependency, npm allows declaring a dependency as a ["peer
+dependency"](https://nodejs.org/es/blog/npm/peer-dependencies). This means
+consumers are expected to **directly** install all transitive dependencies. In
+the *aws-ecs* case, this means that instead of declaring a dependency only on
+*aws-ecs*, users would have to declare a dependency on all 23 transitive modules
+that *aws-ecs* depend on.
+
+This is an unacceptable user experience with our current tools.
+
+See the [appendix](#appendix-cdk-and-npm-dependencies) written by
+[@rix0rrr](https://github.com/rix0rrr) which describes this problem in detail.
+
+### Poor ergonomics due to too many modules
+
+In addition to the peer npm dependency issue described above, due to the fact
+that CDK is "hyper modular", in all languages, we have poor ergonomics when it
+comes to declaring dependencies. Since users are required to explicitly install
+a module for each service they use, even simple projects end up with dozens of
+direct CDK dependencies.
+
+## Experimental Modules
+
+In some cases, a CDK module can include both stable and experimental constructs.
+For example, a module like **@aws-cdk/aws-eks** includes stable CFN resources (L1s)
+but also a variety of experimental L2 classes.
+
+Our API stability guarantees and adherence to semantic versioning only applies
+to STABLE APIs and not to EXPERIMENTAL APIs.
+
+This situation is confusing to users. It is not uncommon for users to use
+experimental constructs without being aware and then be surprised by breaking
+changes in minor versions.
+
+The current approach makes it hard for us to specify the stability at the module
+level since the L1s are stable by definition, but all high-level constructs are
+still experimental.
```

Comment:
This section is just describing a problem with a solution in mind where the solution will be breaking.

In particular, I think the solution is to move classes that have experimental APIs to a separate package, which means that if you want them you need to update your source to point to the new package, and hence a breaking change.

Reply:
For the record, I think we're going to bring pain onto ourselves if we want to split off experimental APIs into a separate package. Here are some thought exercises:

* What if we want to add an experimental method to an otherwise stable class? Do we copy the complete class?
* What if we want to add an experimental property to an otherwise stable props struct? Do we copy the complete class?
* What if we want to provide a stable API that depends on an experimental interface? (Remember, we can easily do this and there is no reason we shouldn't: the breakage caused by any changes to the experimental API will be detected by our complete in-tree build and corrected by the change maker, leaving to trace of it to downstream consumers) We won't be able to do that as we'll get cyclic dependencies.
* What if we stabilize an experimental class? Do we move it from the experimental package to the stable package, thereby breaking everyone in the process?

The better solution is to vend 2 different flavors of CDK from a single codebase:

* `stable` APIs only
* `stable+experimental` APIs

This will solve all issues mentioned above:

* Add experimental features to stable classes: just will not appear in the `stable` flavor, WILL appear in the `s+e` flavor.
* Depend on experimental APIs from stable: will just work, experimental API is just considered *private* in the `s` flavor.
* Stabilizing a class: `s+e` flavor will continue to work as usual, class now appears in `s` flavor.

Reply:
@nija-at added a sentence to clarify why this is a breaking change
@rix0rrr I will take your points into address when I expand the experimental modules section